From 8de2c2176210a5d438108214555c906ef238ec94 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 21 Feb 2024 10:33:33 +0100
Subject: [PATCH 0624/1695] drivers: stm32_rifsc: differentiate STM32MP2x RIFSC
 implementations

Differentiate STM32MP21 from STM32MP23 and STM32MP25 implementations
of the RIFSC as the STM32MP21 embeds different peripherals.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 346af84508bfd8e222f4229e9f2c5dd4cb234dc1 drivers: stm32_rifsc: differentiate STM32MP2x RIFSC implementations]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I55657b02414e639fd89693b7900d12aca8fec66b
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/361574
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 106 ++++++++++++++++++++++++++--
 1 file changed, 101 insertions(+), 5 deletions(-)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index 9b506f71c..362323ca2 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -6,8 +6,7 @@
 #include <drivers/stm32_firewall.h>
 #include <drivers/stm32_rif.h>
 #include <drivers/stm32_rifsc.h>
-#include <dt-bindings/soc/stm32mp25-rif.h>
-#include <dt-bindings/soc/stm32mp25-rifsc.h>
+#include <drivers/stm32mp_dt_bindings.h>
 #include <io.h>
 #include <kernel/boot.h>
 #include <kernel/dt.h>
@@ -786,8 +785,9 @@ out:
 	return res;
 }
 
+#if (defined(CFG_STM32MP25) || defined(CFG_STM32MP23))
 static struct stm32_firewall_reg
-rifsc_compat_mp25[STM32MP25_RIFSC_IAC_ID + 1] = {
+rifsc_compat_ids[STM32MP25_RIFSC_IAC_ID + 1] = {
 	[STM32MP25_RIFSC_TIM1_ID] = { 0x40200000 },
 	[STM32MP25_RIFSC_TIM2_ID] = { 0x40000000 },
 	[STM32MP25_RIFSC_TIM3_ID] = { 0x40010000 },
@@ -906,11 +906,107 @@ rifsc_compat_mp25[STM32MP25_RIFSC_IAC_ID + 1] = {
 	[STM32MP25_RIFSC_OTFDEC2_ID] = { 0x40460000 },
 	[STM32MP25_RIFSC_IAC_ID] = { 0x42090000 },
 };
+#endif /* CFG_STM32MP25 */
+
+#if defined(CFG_STM32MP21)
+static struct stm32_firewall_reg
+rifsc_compat_ids[STM32MP21_RIFSC_IAC_ID + 1] = {
+	[STM32MP21_RIFSC_TIM1_ID] = { 0x40200000 },
+	[STM32MP21_RIFSC_TIM2_ID] = { 0x40000000 },
+	[STM32MP21_RIFSC_TIM3_ID] = { 0x40010000 },
+	[STM32MP21_RIFSC_TIM4_ID] = { 0x40020000 },
+	[STM32MP21_RIFSC_TIM5_ID] = { 0x40030000 },
+	[STM32MP21_RIFSC_TIM6_ID] = { 0x40040000 },
+	[STM32MP21_RIFSC_TIM7_ID] = { 0x40050000 },
+	[STM32MP21_RIFSC_TIM8_ID] = { 0x40210000 },
+	[STM32MP21_RIFSC_TIM10_ID] = { 0x401C0000 },
+	[STM32MP21_RIFSC_TIM11_ID] = { 0x401D0000 },
+	[STM32MP21_RIFSC_TIM12_ID] = { 0x40060000 },
+	[STM32MP21_RIFSC_TIM13_ID] = { 0x40070000 },
+	[STM32MP21_RIFSC_TIM14_ID] = { 0x40080000 },
+	[STM32MP21_RIFSC_TIM15_ID] = { 0x40250000 },
+	[STM32MP21_RIFSC_TIM16_ID] = { 0x40260000 },
+	[STM32MP21_RIFSC_TIM17_ID] = { 0x40270000 },
+	[STM32MP21_RIFSC_LPTIM1_ID] = { 0x40090000 },
+	[STM32MP21_RIFSC_LPTIM2_ID] = { 0x400A0000 },
+	[STM32MP21_RIFSC_LPTIM3_ID] = { 0x46050000 },
+	[STM32MP21_RIFSC_LPTIM4_ID] = { 0x46060000 },
+	[STM32MP21_RIFSC_LPTIM5_ID] = { 0x46070000 },
+	[STM32MP21_RIFSC_SPI1_ID] = { 0x40230000 },
+	[STM32MP21_RIFSC_SPI2_ID] = { 0x400B0000 },
+	[STM32MP21_RIFSC_SPI3_ID] = { 0x400C0000 },
+	[STM32MP21_RIFSC_SPI4_ID] = { 0x40240000 },
+	[STM32MP21_RIFSC_SPI5_ID] = { 0x40280000 },
+	[STM32MP21_RIFSC_SPI6_ID] = { 0x40350000 },
+	[STM32MP21_RIFSC_SPDIFRX_ID] = { 0x400D0000 },
+	[STM32MP21_RIFSC_USART1_ID] = { 0x40330000 },
+	[STM32MP21_RIFSC_USART2_ID] = { 0x400E0000 },
+	[STM32MP21_RIFSC_USART3_ID] = { 0x400F0000 },
+	[STM32MP21_RIFSC_UART4_ID] = { 0x40100000 },
+	[STM32MP21_RIFSC_UART5_ID] = { 0x40110000 },
+	[STM32MP21_RIFSC_USART6_ID] = { 0x40220000 },
+	[STM32MP21_RIFSC_UART7_ID] = { 0x40370000 },
+	[STM32MP21_RIFSC_LPUART1_ID] = { 0x46030000 },
+	[STM32MP21_RIFSC_I2C1_ID] = { 0x40170000 },
+	[STM32MP21_RIFSC_I2C2_ID] = { 0x40180000 },
+	[STM32MP21_RIFSC_I2C3_ID] = { 0x46040000 },
+	[STM32MP21_RIFSC_SAI1_ID] = { 0x40290000 },
+	[STM32MP21_RIFSC_SAI2_ID] = { 0x402A0000 },
+	[STM32MP21_RIFSC_SAI3_ID] = { 0x402B0000 },
+	[STM32MP21_RIFSC_SAI4_ID] = { 0x40340000},
+	[STM32MP21_RIFSC_MDF1_ID] = { 0x404D0000 },
+	[STM32MP21_RIFSC_FDCAN_ID] = { 0xFFFFFFFF },
+	[STM32MP21_RIFSC_HDP_ID] = { 0x44090000 },
+	[STM32MP21_RIFSC_ADC1_ID] = { 0x404E0000 },
+	[STM32MP21_RIFSC_ADC2_ID] = { 0x404F0000 },
+	[STM32MP21_RIFSC_ETH1_ID] = { 0x482C0000 },
+	[STM32MP21_RIFSC_ETH2_ID] = { 0x482D0000 },
+	[STM32MP21_RIFSC_USBH_ID] = { 0xFFFFFFFF },
+	[STM32MP21_RIFSC_OTG_HS_ID] = { 0x48300000 },
+	[STM32MP21_RIFSC_STGEN_ID] = { 0x48080000 },
+	[STM32MP21_RIFSC_OCTOSPI1_ID] = { 0x40430000 },
+	[STM32MP21_RIFSC_SDMMC1_ID] = { 0x48220000 },
+	[STM32MP21_RIFSC_SDMMC2_ID] = { 0x48230000 },
+	[STM32MP21_RIFSC_SDMMC3_ID] = { 0x48240000 },
+	[STM32MP21_RIFSC_LTDC_CMN_ID] = { 0x48010000 },
+	[STM32MP21_RIFSC_CSI_ID] = { 0x48020000 },
+	[STM32MP21_RIFSC_DCMIPP_ID] = { 0x48030000 },
+	[STM32MP21_RIFSC_DCMI_PSSI_ID] = { 0xFFFFFFFF },
+	[STM32MP21_RIFSC_CCB_ID] = { 0x42037C00 },
+	[STM32MP21_RIFSC_RNG1_ID] = { 0x42030800 },
+	[STM32MP21_RIFSC_RNG2_ID] = { 0x42020000 },
+	[STM32MP21_RIFSC_PKA_ID] = { 0x42032000 },
+	[STM32MP21_RIFSC_SAES_ID] = { 0x42030c00 },
+	[STM32MP21_RIFSC_HASH1_ID] = { 0x42030400 },
+	[STM32MP21_RIFSC_HASH2_ID] = { 0x42010000 },
+	[STM32MP21_RIFSC_CRYP1_ID] = { 0x42030000 },
+	[STM32MP21_RIFSC_CRYP2_ID] = { 0x42040000 },
+	[STM32MP21_RIFSC_IWDG1_ID] = { 0x44010000 },
+	[STM32MP21_RIFSC_IWDG2_ID] = { 0x44020000 },
+	[STM32MP21_RIFSC_IWDG3_ID] = { 0x44030000 },
+	[STM32MP21_RIFSC_IWDG4_ID] = { 0x44040000 },
+	[STM32MP21_RIFSC_WWDG1_ID] = { 0x44050000 },
+	[STM32MP21_RIFSC_VREFBUF_ID] = { 0x44060000 },
+	[STM32MP21_RIFSC_DTS_ID] = { 0x44070000 },
+	[STM32MP21_RIFSC_RAMCFG_ID] = { 0x42070000 },
+	[STM32MP21_RIFSC_CRC_ID] = { 0x404C0000 },
+	[STM32MP21_RIFSC_SERC_ID] = { 0x44080000 },
+	[STM32MP21_RIFSC_I3C1_ID] = { 0x40190000 },
+	[STM32MP21_RIFSC_I3C2_ID] = { 0x401A0000 },
+	[STM32MP21_RIFSC_I3C3_ID] = { 0x401B0000 },
+	[STM32MP21_RIFSC_ICACHE_DCACHE_ID] = { 0x40470000 },
+	[STM32MP21_RIFSC_LTDC_L0L1_ID] = { 0xFFFFFFFF },
+	[STM32MP21_RIFSC_LTDC_L2_ID] = { 0xFFFFFFFF },
+	[STM32MP21_RIFSC_LTDC_ROT_ID] = { 0xFFFFFFFF },
+	[STM32MP21_RIFSC_OTFDEC1_ID] = { 0x40450000 },
+	[STM32MP21_RIFSC_IAC_ID] = { 0x42090000 },
+};
+#endif /* CFG_STM32MP21 */
 
 static const struct stm32_firewall_compat rifsc_compat[] = {
 	{
-		.reg = rifsc_compat_mp25,
-		.compat_size = ARRAY_SIZE(rifsc_compat_mp25)
+		.reg = rifsc_compat_ids,
+		.compat_size = ARRAY_SIZE(rifsc_compat_ids)
 	},
 };
 
-- 
2.47.2

